presto-main/src/main/java/com/facebook/presto/operator/scalar/ArrayUnionFunction.java

```diff
@@ -91,4 +103,28 @@ private static void appendBigintArray(Block array, AtomicBoolean containsNull, L
             }
         }
     }
+
+    private static void appendTypedArray(Block array, TypedSet typedSet)
+    {
+        for (int i = 0; i < array.getPositionCount(); i++) {
+            typedSet.add(array, i);
+        }
+    }
+
+    private static boolean[] getDistinctBooleanValues(Block block)
+    {
+        boolean[] results = new boolean[3];
+
+        for (int i = 0; i < block.getPositionCount(); i++) {
+            if (block.isNull(i)) {
+                results[0] = true;
+            }
+            else {
+                boolean value = BOOLEAN.getBoolean(block, i);
+                results[1] |= value;
+                results[2] |= !value;
```

Comment:
Is this better than

```
else if (BOOLEAN.getBoolean(block, i)) {
    results[1] = true;
}
else {
    results[2] = true;
}
```

?

Reply:
> Is this better than

The original form is certainly less readable, so unless there's a measurable and significant impact to performance, let's go with the readable version.

Reply:
Using this way the performance is almost 3x worse.
Benchmark                        (type)  Mode  Cnt  Score   Error  Units
BenchmarkArrayUnion.arrayUnion  BOOLEAN  avgt   20  9.097 ± 0.496  ns/op

The branching in this case is on every value and is quite costly. It stalls the CPU pipeline and we should try to minimize it when possible.

Reply:
I can't reproduce those results. These are the measurements I get for three variants I tried (on my macbook pro, 2.9 GHz Intel Core i9, Java 1.8.0_181):

1. The version I suggested above: 1.596 ± 0.053  ns/op
2. Similar to what I suggested above but using |= instead of if: 1.034 ± 0.052  ns/op
3. This PR: 2.740 ± 0.088  ns/op

How do you draw the conclusion that branching is stalling the CPU pipeline? Do you have a capture from perf that indicates so?

BTW, the benchmark in this PR is very biased. It doesn't test for different lengths of arrays, doesn't test when nulls are present, doesn't test for all true, all false, etc.

Reply:
You can make it even faster (0.782 ± 0.087 ns/op) by having a list of pre-computed blocks for all the 8 combinations of answers this method produces. Not that it matters much, especially for the additional complexity -- we're talking sub-nanosecond per element gains.

Reply:
I reran the benchmark with 10, 100, 1000 elements in the array. The difference is reproducible if booleanUnion is calling getDistinctBooleanValues on my i7 processor on both JDK8 and 10

Calling getDistinctBooleanValues function, using |= (This PR)
Benchmark                       (ARRAY_SIZE)   (type)  Mode  Cnt    Score    Error  Units
BenchmarkArrayUnion.arrayUnion            10  BOOLEAN  avgt   20   10.823 ±  3.306  ms/op
BenchmarkArrayUnion.arrayUnion           100  BOOLEAN  avgt   20   28.701 ±  0.789  ms/op
BenchmarkArrayUnion.arrayUnion          1000  BOOLEAN  avgt   20  259.446 ± 64.891  ms/op

Calling getDistinctBooleanValues function, using if else
Benchmark                       (ARRAY_SIZE)   (type)  Mode  Cnt    Score    Error  Units
BenchmarkArrayUnion.arrayUnion            10  BOOLEAN  avgt   20   14.123 ±  0.441  ms/op
BenchmarkArrayUnion.arrayUnion           100  BOOLEAN  avgt   20   86.883 ±  4.807  ms/op
BenchmarkArrayUnion.arrayUnion          1000  BOOLEAN  avgt   20  869.110 ± 21.275  ms/op

The two getDistinctBooleanValues() functions are as follows and this is the only difference between the two runs:

```
private static boolean[] getDistinctBooleanValues(Block block)
    {
        boolean[] results = new boolean[3];

        for (int i = 0; i < block.getPositionCount(); i++) {
            if (block.isNull(i)) {
                results[0] = true;
            }
            else {
                boolean value = BOOLEAN.getBoolean(block, i);
                results[1] |= value;
                results[2] |= !value;
            }
        }
        return results;
    }
```

   ```
 private static boolean[] getDistinctBooleanValues(Block block)
    {
        boolean[] results = new boolean[3];

        for (int i = 0; i < block.getPositionCount(); i++) {
            if (block.isNull(i)) {
                results[0] = true;
            }
            else if (BOOLEAN.getBoolean(block, i)) {
                results[1] = true;
            }
            else {
                results[2] = true;
            }
        }
        return results;
    }
```
However as you said, writing the code in the inlined way as shown in your code below doesn't reproduce the performance difference.  

Originally I thought the difference shown in my above benchmark was because of the branch mis-prediction because the only difference between the two getDistinctBooleanValues implementation was the use of |=. But I looked at the assembly and didn't see any difference. In both cases the functions got inlined. So this may not be a branch misprediction issue. In this case the if-else are writing to two different variables instead of one and they tend to converge to true on the given workload, and the code block sizes are very small, so branch misprediction may not have such high impact. Also skylake has much better branch predictor which may also be the reason there was no difference on both i7 and i9 CPUs. Anyways, I get a profile on it to understand why there is the discrepancy in my runs.